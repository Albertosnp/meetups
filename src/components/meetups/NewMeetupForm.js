import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { addNewMeetup } from 'stores/meetups.slice';
import Card from '../ui/Card';
import classes from './NewMeetupForm.module.css';

const INITIAL_FORM_STATE = {
  title: '',
  image: '',
  address: '',
  description: '',
};

export default function NewMeetupForm() {
  const dispatch = useDispatch();
  const [formValues, setFormValues] = useState(INITIAL_FORM_STATE);

  function submitHandler(event) {
    event.preventDefault();

    dispatch(addNewMeetup(formValues));
  }

  const handleChange = ({ target }) => {
    setFormValues((prev) => ({
      ...prev,
      [target.name]: target.value,
    }));
  };

  return (
    <Card>
      <form className={classes.form} onSubmit={submitHandler}>
        <div className={classes.control}>
          <label htmlFor="title">Meetup Title</label>
          <input
            type="text"
            required
            id="title"
            name="title"
            value={formValues.title}
            onChange={handleChange}
          />
        </div>
        <div className={classes.control}>
          <label htmlFor="image">Meetup Image</label>
          <input
            type="url"
            required
            id="image"
            name="image"
            value={formValues.image}
            onChange={handleChange}
          />
        </div>
        <div className={classes.control}>
          <label htmlFor="address">Address</label>
          <input
            type="text"
            required
            id="address"
            name="address"
            value={formValues.address}
            onChange={handleChange}
          />
        </div>
        <div className={classes.control}>
          <label htmlFor="description">Description</label>
          <textarea
            id="description"
            required
            rows="5"
            name="description"
            value={formValues.description}
            onChange={handleChange}
          />
        </div>
        <div className={classes.actions}>
          <button>Add Meetup</button>
        </div>
      </form>
    </Card>
  );
}
