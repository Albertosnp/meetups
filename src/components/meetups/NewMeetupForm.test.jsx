import { configure, fireEvent, render, screen, waitFor } from '@testing-library/react';
import ReactDOM from 'react-dom';
import * as ReactRedux from 'react-redux';

import { mockComponent } from '../../utils/test-helpers/mockComponent';
import NewMeetupForm from './NewMeetupForm';
import { addNewMeetup } from '../../stores/meetups.slice';
import userEvent from '@testing-library/user-event';

configure({ testIdAttribute: 'data-cy' });

jest.mock('../ui/Card', () => props => mockComponent('Card', props));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(() => jest.fn(() => null))
}));

jest.mock('../../stores/meetups.slice', () => ({
  ...jest.requireActual('../../stores/meetups.slice'),
  addNewMeetup: jest.fn(() => jest.fn(() => null))
}));

describe('NewMeetupForm', () => {
  const defaultProps = {};
  const getComponent = props => <NewMeetupForm {...props} />;
  const setup = (props = defaultProps) => render(getComponent(props));

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(getComponent(), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot with full props', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should change text of title input', async () => {
    setup();
    await waitFor(() => {
      const inputElement = screen.getByRole('textbox', { name: 'Meetup Title' });
      fireEvent.change(inputElement, { target: { value: 'New meetup test' } });
      expect(inputElement.value).toBe('New meetup test');
    });
  });

  it('Should call addNewMeetup when send all fields', async () => {
    const dispatch = jest.fn(() => null);
    ReactRedux.useDispatch.mockReturnValue(dispatch);
    setup();

    const inputElement = screen.getByRole('textbox', { name: 'Meetup Title' });
    userEvent.type(inputElement, 'New meetup test');

    const inputElement2 = screen.getByRole('textbox', { name: 'Meetup Image' });
    userEvent.type(inputElement2, 'Meetup Image');

    const inputElement3 = screen.getByRole('textbox', { name: 'Address' });
    userEvent.type(inputElement3, 'New address');

    const inputElement4 = screen.getByRole('textbox', { name: 'Description' });
    userEvent.type(inputElement4, 'New description');

    await waitFor(() => {
      const button = screen.getByText('Add Meetup');
      userEvent.click(button);
      expect(addNewMeetup).toHaveBeenCalledTimes(1);
      expect(addNewMeetup).toHaveBeenCalledWith({
        address: 'New address',
        description: 'New description',
        image: 'MeetupImage',
        title: 'New meetup test'
      });
    });
  });
});
