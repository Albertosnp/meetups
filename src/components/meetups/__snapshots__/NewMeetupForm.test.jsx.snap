// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewMeetupForm Should match snapshot with full props 1`] = `
<div>
  <code
    aria-label="Card"
    name="Card"
    props="{}"
  >
    <form
      class="form"
    >
      <div
        class="control"
      >
        <label
          for="title"
        >
          Meetup Title
        </label>
        <input
          id="title"
          name="title"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="control"
      >
        <label
          for="image"
        >
          Meetup Image
        </label>
        <input
          id="image"
          name="image"
          required=""
          type="url"
          value=""
        />
      </div>
      <div
        class="control"
      >
        <label
          for="address"
        >
          Address
        </label>
        <input
          id="address"
          name="address"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="control"
      >
        <label
          for="description"
        >
          Description
        </label>
        <textarea
          id="description"
          name="description"
          required=""
          rows="5"
        />
      </div>
      <div
        class="actions"
      >
        <button>
          Add Meetup
        </button>
      </div>
    </form>
  </code>
</div>
`;
